### PR TITLE
Restore sound

### PR DESCRIPTION
**Pull request by kchadha**
### Resolves

Towards restoring a deleted sound.

### Proposed Changes

Deleting a sound returns a function to restore the deleted sound.

### Reason for Changes

Restore sound feature.

### Test Coverage

Added new unit test for vm.deleteSound and updated a unit test for rendered-target.deleteSound.

### Related PRs
LLK/scratch-gui#2935

### CODE REVIEW

File: src/virtual-machine.js [hunk 1]
```diff
@@ -675,9 +675,20 @@ class VirtualMachine extends EventEmitter {
     /**
      * Delete a sound from the current editing target.
      * @param {int} soundIndex - the index of the sound to be removed.
+     * @return {?Function} A function to restore the sound that was deleted,
```
Comment (kchadha) [inline src/virtual-machine.js:678]: Should the `F` in function be capital or not? I've seen it both ways in this same file... O.o

File: src/sprites/rendered-target.js [hunk 2]
```diff
@@ -585,12 +585,17 @@ class RenderedTarget extends Target {
...
     deleteSound (index) {
-        this.sprite.sounds = this.sprite.sounds
-            .slice(0, index)
-            .concat(this.sprite.sounds.slice(index + 1));
+        // Make sure the sound index is not out of bounds
+        if (index < 0 || index >= this.sprite.sounds.length) {
+            return null;
+        }
+        // Delete the sound at the given index
+        const deletedSound = this.sprite.sounds.splice(index, 1)[0];
```
Comment (kchadha) [inline src/sprites/rendered-target.js:596]: Note the switch to splice here. I wanted to make sure this is okay, or check if there was any particular reason why we weren't using splice before...

File: test/unit/sprites_rendered-target.js [hunk 3]
```diff
@@ -211,8 +211,10 @@ test('deleteSound', t => {
     const renderer = new FakeRenderer();
     a.renderer = renderer;
 
-    a.deleteSound(0);
+    const firstDeleted = a.deleteSound(0);
     t.deepEqual(a.sprite.sounds, [o2, o3]);
+    t.type(firstDeleted, 'object');
+    t.equals(firstDeleted.id, 1);
```
Comment (rschamp) [inline test/unit/sprites_rendered-target.js]: Can we not do a similar `t.deepEqual` for the returned sound?

File: src/sprites/rendered-target.js (hunk 2 above)
Comment (rschamp) [reply src/sprites/rendered-target.js:596]: I don't see anything about using `slice` over `splice`, so might just be an improvement!

File: src/virtual-machine.js (hunk 1 above)
Comment (rschamp) [reply src/virtual-machine.js:678]: Looking at https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#function-type it looks like lowercase...

**Review by rschamp — APPROVED** [review]
Looks good!

File: src/virtual-machine.js (hunk 1 above)
Comment (kchadha) [reply src/virtual-machine.js:678]: According to this thread I found earlier, `function` with a lowercase `f` doesn't exist...?
https://github.com/Microsoft/TypeScript/issues/10946#issuecomment-248084184

scratch-blocks seems to also use both lowercase and uppercase, so I'm just going to stick with uppercase.

File: test/unit/sprites_rendered-target.js (hunk 3 above)
Comment (kchadha) [reply test/unit/sprites_rendered-target.js]: Done!

**Pull request merged**